Improvement in satask

ShubhamKJha opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Built over  #11789.
A lot of time in the execution of `satask` is wasted in unnecessary creation of `And` and `Or` objects. This PR tries to remove these intermediate steps.
The current results in performance:

| Tests             | This PR         | This PR (most recent)    |  Master        |
| ------------- | -------------   | -------------  | -------------|
| test_satask    | 2.39 s   | 1.06 s | 36.26 s|
| assumptions\tests | 16.74 | 7.61 s     | 127.21 |

#### Brief description of what is fixed or changed


#### Other comments
Todos:
* [x]  Fix failing tests in `ask` module.
* [ ]  Implement early encoding of `Literal`s


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* assumptions
  * Performance of ask and satask is now improved.
  * Implemented new `cnf.py` to handle logical computations using low-level Python built-ins.
<!-- END RELEASE NOTES -->

rlamy committed:
Start using a KB in dpll_satisfiable()

rlamy committed:
Simplify get_all_relevant_facts()

rlamy committed:
More get_relevant_facts() simplification

rlamy committed:
Combine assumptions and context earlier

rlamy committed:
Return relevant_facts as a set

rlamy committed:
Bypass dpll_satisfiable() to remove yet another costly And instantiation

(note: tautology corner cases handled incorrectly)

rlamy committed:
CNF.from_prop()

rlamy committed:
Simplify get_relevant_facts() again

rlamy committed:
Keep known_facts in CNF form rather than reconverting them

rlamy committed:
Reify encoding step for SAT solver

rlamy committed:
Extract know facts handling from add_relevant_facts()

rlamy committed:
Rename KB to EncodedCNF and make the encoding explicit

rlamy committed:
Give EncodedCNF a proper __init__ and turn the old one into an alternate constructor

rlamy committed:
SAT-encode a bit earlier

rlamy committed:
Bypass very expensive 'known_facts_CNF.rcall(expr)'

ShubhamKJha committed:
fix merge conflicts

ShubhamKJha committed:
Initial commit

ShubhamKJha committed:
Deleted extra function and changed ask_generated

ShubhamKJha committed:
Removed redundant whitespace

ShubhamKJha committed:
Removed redundant whitespace

ShubhamKJha committed:
Fix for failing tests

ShubhamKJha committed:
Made the code of dpll_satisfiable to work with standard input

ShubhamKJha committed:
Merge branch 'speed_up_cnf_3' of https://github.com/ShubhamKJha/sympy into speed_up_cnf_3

oscargus commented:
Looks promising! However, since you base it on #11789 you should include those commits here (with correct authorship). Not sure how to go that in an easy way though, but maybe someone else knows.

oscargus commented:
Not sure that all of these are from assumptions, but many are:

```
___________________________ unexpectedly fast tests ____________________________
...
test_pow_pos_neg - Took 0.588 seconds
test_zero - Took 1.123 seconds
test_negative - Took 1.177 seconds
test_bounded1 - Took 1.498 seconds
test_hermitian - Took 1.518 seconds
test_positive - Took 1.681 seconds
test_pow2 - Took 1.867 seconds
test_pow4 - Took 1.883 seconds
test_bounded3 - Took 1.955 seconds
test_real_pow - Took 2.641 seconds
test_pow3 - Took 2.760 seconds
test_bounded2a - Took 3.033 seconds
test_even_query - Took 3.134 seconds
test_issue_11238 - Took 4.158 seconds
test_odd_query - Took 4.445 seconds
```

Nice!

ShubhamKJha committed:
Sped up the ask module

ShubhamKJha committed:
Some changes

oscarbenjamin commented:
@ShubhamKJha this looks like it will be useful but you need to fix the attribution of the commits.

The way I would do this is to start a new branch as a copy of the old PR (with previous author information intact), rebase that on current master, then apply my changes on top.

ShubhamKJha committed:
An attempt to make sympify faster for sympy objects

ShubhamKJha committed:
to_CNF a faster version of to_cnf made to work with CNF objects directly

ShubhamKJha committed:
Minor fixes and commented a significant part of ask function's code which seems can be improved

ShubhamKJha commented:
Hi @oscarbenjamin and @oscargus, I didn't start from that PR, so I faced many merge conflicts on a local branch I tried. Also, since both PRs have some content in same files, the code is getting merged unexpectedly (it may need more than merging). Anyhow, this is a WIP branch and I would try to give proper attribution to the original commits once things work out. I would then open another PR over this.

ShubhamKJha committed:
Used class.__name__ instead of class for speed

asmeurer commented:
I think the simplest way would be to take this branch and rebase it on top of Ronan's. Another option, if you want to avoid fixing merge conflicts for each commit, would be to merge Ronan's branch into this one, but with the merge conflict resolution set to keep the changes here (I think it's `git merge --strategy ours`). That way, Ronan's commits will be in the history, but you won't have to deal with actually managing the differences between the two with git. If you go with the latter approach, just explain what is going on in the merge commit message.

ShubhamKJha committed:
Some changes to make to_CNF faster

ShubhamKJha committed:
Attempt to reduce time taken

ShubhamKJha committed:
Created a function to_NNF to first convert the boolean expression into NNF and created a function Distribute that will convert the NNF expression to Conjunctive Normal Form

ShubhamKJha committed:
Some minor issues

ShubhamKJha committed:
Added proper attribution to the original commits made in PR #11789 by @rlamy.

ShubhamKJha committed:
Created file sympy/assumptions/CNF.py to hold the code related to the generation and handling of CNF objects

ShubhamKJha committed:
Fixed some failing tests

ShubhamKJha commented:
Followed the second method Aaron suggested.

Some fast tests related to this (tested locally):
```
___________________________ unexpectedly fast tests ____________________________
test_pow_pos_neg - Took 0.370 seconds
test_negative - Took 0.515 seconds
test_zero - Took 0.571 seconds
test_bounded1 - Took 0.586 seconds
test_positive - Took 0.703 seconds
test_hermitian - Took 0.780 seconds
test_bounded3 - Took 0.871 seconds
test_pow4 - Took 0.887 seconds
test_pow2 - Took 1.103 seconds
test_bounded2a - Took 1.165 seconds
test_even_query - Took 1.172 seconds
test_pow3 - Took 1.272 seconds
test_real_pow - Took 1.654 seconds
test_odd_query - Took 2.077 seconds
test_bounded2b - Took 2.090 seconds
test_imaginary - Took 2.544 seconds

```

ShubhamKJha committed:
Fixed some errors in satisfiable

ShubhamKJha committed:
Fixed code in satisfiable

oscarbenjamin commented on sympy/assumptions/ask.py:
```diff
@@ -8,9 +8,10 @@
+                                 BooleanFunction, BooleanAtom)
 from sympy.logic.inference import satisfiable
 from sympy.utilities.decorator import memoize_property
+from sympy.assumptions.CNF import CNF, EncodedCNF, Literal
```
The module should have a lower case name.

asmeurer replied:
Yes please use lowercase names for modules. Uppercase file names can cause issues on systems with case insensitive filesystems (like macOS or Windows).

oscarbenjamin commented on sympy/assumptions/ask.py:
```diff
@@ -1211,6 +1213,30 @@ def _extract_facts(expr, symbol, check_reversed_rel=True):
+
+            if isinstance(symbol, Relational):
+                if check_reversed_rel:
+                    rev = _extract_all_facts(expr, symbol.reversed, False)
```
This recursive call seems unnecessary. Can you not just change it above to:
```
if literal.lit.arg in (symbol, symbol.reversed)
```

oscarbenjamin commented on sympy/core/sympify.py:
```diff
@@ -263,19 +263,16 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
+        if a.__sympy__:
             return a
-    except TypeError: # Type of a is unhashable
+    except AttributeError:  # a is not a sympy object
```
Please don't catch AttributeError as it can hide bugs. It's better to do
```
    attr = getattr(a, '__sympy__', None)
    if attr is not None:
```
(like you do below for `__class__`)

oscarbenjamin commented on sympy/core/sympify.py:
```diff
@@ -263,19 +263,16 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
+        cls = type(a)  # Probably an old-style class
     try:
         return converter[cls](a)
     except KeyError:
```
While you're here that KeyError is being caught to broadly. It should have been written something like:
```
    try:
        c = converter[cls]
    except KeyError:
        # whatever goes here
    else:
        return c(a) # Don't mask KeyErrors leaking from this function call
```
Better still though don't catch KeyError (use get instead):
```
    c = converter.get(cls, None)
    if c is None:
```

asmeurer commented:
Can you explain the purpose of the changes to sympify.

ShubhamKJha commented:
`sympify` is called by many constructors and it is applied on SymPy objects as well. The changes above reduces execution time to about one-third for SymPy objects.

asmeurer commented:
What exactly does the change do? I don't see any commit messages about it. If it changes the behavior of sympify it should be documented in the release notes.

ShubhamKJha committed:
Changed module name to lower case

ShubhamKJha committed:
Fixed the failing tests in ask module. Created function _extract_all_facts to mimic the _extract_facts function but with CNF objects and changed the code in ask function accordingly.

ShubhamKJha committed:
Minor changes for ask_generated

oscarbenjamin commented:
This is a lot faster so this PR will be a significant improvement I think. I reprofiled this. The profile I used in #17066 wasn't relevant any more because it was so much faster so this time I used all the tests in sympy/assumptions/tests.

To reproduce this profile install gprof2dot and graphviz and run
```
python -m cProfile -o output.pstats bin/test sympy/assumptions/tests/test_*.py --no-subprocess
gprof2dot -f pstats output.pstats -n3 -e1 | dot -Tpng -o output.png
```
The result:
![output](https://user-images.githubusercontent.com/1159732/61589957-39429b80-aba9-11e9-905b-8701b58d1ee6.png)

oscarbenjamin commented:
Testing some simple queries like `ask(Q.real(e), ...)` vs e.is_real suggests that this makes the new assumptions gives roughly
```
new assumptions (master): 45 seconds
new assumptions (this pr): 3 seconds
old assumptions:           1 second
```
So this PR makes the new assumptions a lot faster. They are still maybe 3 times slower than the old assumptions but I think that's a small enough difference that you could actually contemplate using the new assumptions in core.

ShubhamKJha committed:
Updated code in sympify.py module, reduced the need of catching AttributeError, TypeError and KeyError. Update code in core.py made __sympy__ a property for SymPy objects so as to avoid changing its value.

ShubhamKJha commented:
ping @asmeurer, @oscarbenjamin, @jksuom. I think this PR can be reviewed now. 

The failing tests in `ask` are fixed now and I have tried early encoding for `Literal`s. However, the literals are still encoded again in `EncodedCNF`. This has remained because of the hack used by Ronan for speeding up `rcall` applied on `known_facts_cnf`, in `get_all_relevant_facts` during `satask`. I am trying to improve the performance of `rcall` for `CNF` objects so as to remove this hack and the need for `EncodedCNF` at all. But till then we can have this merged since the performance gain is good.

> What exactly does the change do? I don't see any commit messages about it. If it changes the behavior of sympify it should be documented in the release notes.

I have modified the code of `BasicMeta` so it adds a new property `__sympy__` to all the classes and hence all the objects of SymPy. In order to check if the argument is a SymPy object, the code in `sympify` earlier checked in `all_classes` set ( both for argument and it's class). Compared to that checking of a property is faster. Also, `git grep` for `all_classes` suggest that it is used only in `/deprecated/class_registry.py` other than `sympify`. The changes here doesn't modify the behavior of `sympify` just make it faster, hence didn't mention it in the release notes.

> They are still maybe 3 times slower than the old assumptions

For new assumptions, it also depends upon the number of clauses used. With each assumption, it may increase exponentially and hence the execution time can be very varied. 

> So this PR makes the new assumptions a lot faster.

There is still some scope to make it faster.

oscarbenjamin commented:
>  am trying to improve the performance of `rcall` for `CNF` objects so as to remove this hack and the need for `EncodedCNF` at all.

That sounds like a good thing to do.

> But till then we can have this merged since the performance gain is good.

I agree. This is a good change as it is and can be improved in future PRs.

> There is still some scope to make it faster.

We have to be careful here. Removing hacks and cleaning things up is obviously good. If there are some clear speed wins that don't impact on the clarity etc of the code then they are worth doing. Removing the EncodedCNF hack sounds like a good example.

However there is an argument to be made for avoiding over-optimisation at this stage. We don't yet have any data about how realistic usage of the new assumptions looks. We may end up spending a lot of time optimising things that don't turn out to be important. It's possible even that over-optimising for contrived examples now leads to poorer performance in actual usage later.

I would think that once the new assumptions are fast enough to be feasible in the core the next step is to think about how they actually can be incorporated into the core (hopefully incrementally). Actually using them there will expose the cases that need to be made faster.

oscarbenjamin commented on sympy/assumptions/ask.py:
```diff
@@ -1211,6 +1213,30 @@ def _extract_facts(expr, symbol, check_reversed_rel=True):
         return expr.func(*args)
 
 
+def _extract_all_facts(expr, symbol, check_reversed_rel=True):
```
`check_reversed_rel=True` isn't needed any more.

ShubhamKJha commented:
I tried it on a larger example of 
```py
p = random_poly(x, 50, -50, 5) 
ask(Q.positive(p), Q.positive(x))
```
and the overall execution time was `3.117 s`, the amount of time spent (as profiled) are resp.

---
-  satask ( 2.715 s), rest was spent in handlers
- rcall (1.867 s)
- satisfiable ( 0.294 s)
- to_CNF (0.191 s)

---

It clearly shows a major contribution from `rcall`.

oscarbenjamin commented:
I think if you can get rid of the SymPy objects that are created in rcall then that will be good.

It's worth noting though that the old assumptions are slow in this case as well. On my computer I tried a few examples from random_poly and execution time varies from 3-12 seconds for the new assumptions but is always around 10 seconds for the old. (I don't know whether both are equally accurate - they both seem to always give None for these large polys).

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+    if klass == 'ITE':
+        L, M, R = map(CNF.to_CNF, expr.args)
+        return AND(OR(~L, M), OR(L, R))
+
```
`raise NotImplementedError` here if the class isn't a known type.

asmeurer commented:
> I don't know whether both are equally accurate - they both seem to always give None for these large polys

I'm pretty sure the code doesn't do anything smarter than returning True only if all coefficients are positive, or False if they are all negative. 

```py
>>> x = Symbol('x', positive=True)
>>> ((x - 1)**2 + 1).is_positive
True
>>> print(expand((x - 1)**2 + 1).is_positive)
None
>>> ask(Q.positive((x - 1)**2 + 1))
True
>>> print(ask(Q.positive(expand((x - 1)**2 + 1))))
None
```

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+
+
+# Early encoding for literals
+literals_store = dict()
```
This needs to be placed on the CNF object. As it is now, it is a global object that never gets cleared, which will be a memory leak for any long running processes that uses the new assumptions on a lot of SymPy expressions.

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+        k = literals_store.get(lit, None)
+        if k is not None:
+            code = k
+        elif lit in (S.false, False):
```
`False` already compares equal to `S.false`, so you can just use `lit == S.false` here. Unless `0` is supposed to match as well (but unless I am mistaken I think it shouldn't).

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+    literals_count = 0
+
+    def __new__(cls, lit, is_not=False):
+        if type(lit).__name__ == 'Not':
```
Use isinstance here and elsewhere instead of comparing type strings.

ShubhamKJha replied:
Out of everything I tried this is the fastest. `isinstance()` is better when something is going to inherit `Not` and other classes used. I don't think it's going to be a case here.

asmeurer replied:
Does it actually affect the end runtime, though? I tested and they both take on the order of 100 ns.

ShubhamKJha replied:
The advantage comes when the object is not a `Not` object. Though the difference appears to be marginal and I can't say I understand how `isinstance` works but I think it goes into the inheritance of the object also.

oscarbenjamin replied:
Normally `isinstance(obj, cls)` is equivalent to `cls in type(obj).mro()` but it is also possible to override this with metaclasses that define `__instancecheck__`.

There are a couple of problems with what you have here. One is that an unrelated type that happens to have the same name will pass this check because you're just checking the name string. You could fix that by comparing the type objects with `type(lit) is Not` but that also suffers from the other problem which is that the test wouldn't recognise instances of subclasses of Not.

Unless you have a strong (and unusual) reason for doing anything different you should just use isinstance and not think about it. The big gains here are coming from higher-level changes so you should focus on implementing those as cleanly as possible rather than on micro-optimisations like trying to speed up isinstance.

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+        return Literal(self.lit, is_not=is_not)
+
+    def __str__(self):
+        return '%s( %s, %s)' % (type(self).__name__, self.lit, self.is_Not)
```
```suggestion
        return '%s(%s, %s)' % (type(self).__name__, self.lit, self.is_Not)
```

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+
+    def __invert__(self):
+        is_not = not self.is_Not
+        return Literal(self.lit, is_not=is_not)
```
Instead of storing is_not on the object, it would be simpler (and probably faster) to use negative integers for negations.

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+        h = hash((type(self).__name__, + self.arg))
+        return h
+
+    def __lt__(self, other):
```
Where is this used?

ShubhamKJha replied:
In the sorting of Literals. When two same literals are present, it is required for sorting them.

oscarbenjamin replied:
It shouldn't be needed if you provide a sort key (as seems to be defined below). Does this `__lt__` actually make sense mathematically?

ShubhamKJha replied:
In case two literals are equivalent, the sort key doesn't work. `__lt__` is needed for sorting then. Currently, it just compares the encoded form of literals ( e.g. `~Q.positive` is supposedly given an encoding of `-1` and that is compared and that is compared to others').

asmeurer replied:
Why doesn't it work if they are equivalent?

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+        return AND(*[~arg for arg in self._args])
+
+    def __hash__(self):
+        return hash((type(self).__name__, ) + tuple(self.args))
```
```suggestion
        return hash((type(self).__name__,) + tuple(self.args))
```

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+        return sorted(self._args, key=literal_sort_key)
+
+    def __hash__(self):
+        return hash((type(self).__name__, ) + tuple(self.args))
```
```suggestion
        return hash((type(self).__name__,) + tuple(self.args))
```

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+
+    @property
+    def args(self):
+        return sorted(self._args, key=literal_sort_key)
```
Does args get used somewhere? This isn't a Basic object, so it shouldn't need to have it. It seems that anything that uses this should just use the unsorted args (perhaps even store them as a set). My worry is that this sorting could slow things down and seems completely unnecessary.

ShubhamKJha replied:
Sorted `args` is used for equality and printing. I had this while debugging and I think it's better to have it here. The code doesn't use it, instead `_args` is used.

asmeurer replied:
Printing is fine. But I would expect equality to be used in the algorithm somewhere. Can we just store `args` and have it be a set?

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+from itertools import combinations, product
```
Add a docstring to the top of the file explaining that the things in this file are for internal use of the assumptions system only. For everything else, people should use the classes in sympy.logic.

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+from itertools import combinations, product
+
+from sympy import S
+from sympy.logic.boolalg import BooleanFunction, Or, And, Not
```
It might be better to do `from sympy.logic import boolalg` and use `boolalg.Or`, and so on, to avoid confusion with the classes defined here.

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+def to_NNF(expr):
+    """
+    Generates the Negation Normal Form of any boolean expression in terms
+    of AND , OR and Literal objects.
```
```suggestion
    of AND, OR, and Literal objects.
```

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+    if klass == 'Xor':
+        cnfs = []
+        for i in range(0, len(expr.args) + 1, 2):
+            for neg in combinations(expr.args, i):
```
I worry about having this here. Anything using xor will slow things down simply through combinatorial explosion.

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+
+    if klass == 'Equivalent':
+        cnfs = []
+        for a, b in zip(expr.args, expr.args[1:]):
```
You can use zip_longest (from itertools)

```
>>> a = range(4)
>>> from itertools import zip_longest
>>> list(zip_longest(a, a[1:], fillvalue=a[0]))
[(0, 1), (1, 2), (2, 3), (3, 0)]
```

ShubhamKJha replied:
`zip_longest` is not present in Python2, instead the function is `izip_longest`.

oscarbenjamin replied:
This is handled in sympy.compatibility so import it from there

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+    """
+    Class to represent CNF of a Boolean expression.
+    Consists of set of clauses, which themselves are stored as
+    frozenset of Literal objects.
```
```suggestion
    set of Literal objects.
```

asmeurer replied:
Oh I see `clauses` is a set and its elements are frozensets.

ShubhamKJha replied:
A set can't have unhashable types, so `frozenset` is used. I think `frozenset` should be kept.

asmeurer commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+
+    @classmethod
+    def from_prop(cls, prop):
+        res = CNF()
```
```suggestion
        res = cls()
```

ShubhamKJha replied:
Does it have any benefit? AMAIK, this change may break things while inheritance. Though I don't think we may subclass CNF, I would better leave this if there is no benefit.

oscarbenjamin replied:
The change is needed for it to be a proper class method so that it can be overridden by subclasses. It's best not to question things like that because there is no advantage to you in hard-coding the class and it has the potential to break things for someone else down the line.

asmeurer commented on sympy/assumptions/tests/test_query.py:
```diff
@@ -2011,7 +2011,7 @@ def test_composite_assumptions():
     assert ask(Q.real(x), ~(Q.real(x) >> Q.real(y))) is True
 
-
+@XFAIL
```
Why does this fail?

ShubhamKJha replied:
I was trying whether to use handlers first or satask and this test was failing for one of them. 
Will revert it.

asmeurer commented on sympy/assumptions/tests/test_assumptions_2.py:
```diff
@@ -34,7 +34,6 @@ def test_extract_facts():
     assert _extract_facts(a(x) & b(y), x) == a
     assert _extract_facts(a(x) & b(y), y) == b
     assert _extract_facts(a(x) | b(y), x) == None
-    assert _extract_facts(~(a(x) | b(y)), x) == ~a
```
Why was this test removed?

ShubhamKJha replied:
Will revert it.

asmeurer commented on sympy/logic/algorithms/dpll2.py:
```diff
@@ -35,16 +35,18 @@ def dpll_satisfiable(expr, all_models=False):
+        expr = exprs
+
+    # DONE: Check to handle this
+    if {0} in expr.data:
```
What does this mean?

ShubhamKJha replied:
During the encoding, `False` is encoded as `0`. It is used here to test the presence of `False` in `EncodedCNF`.

oscarbenjamin replied:
A more meaningful comment might help

ShubhamKJha replied:
Ohh, that was a TODO I left. I am changing it.

asmeurer commented:
> I have modified the code of BasicMeta so it adds a new property __sympy__ to all the classes and hence all the objects of SymPy. In order to check if the argument is a SymPy object, the code in sympify earlier checked in all_classes set ( both for argument and it's class). Compared to that checking of a property is faster. Also, git grep for all_classes suggest that it is used only in /deprecated/class_registry.py other than sympify. The changes here doesn't modify the behavior of sympify just make it faster, hence didn't mention it in the release notes.

I see. Yes, we want to remove that. It is a lot of work though (see #13059). This is how I changed sympify in that PR: https://github.com/sympy/sympy/pull/13059/files#diff-50fcb76dd49035ed7d6ce34929ba5e0b. How does your change compare to that? Modifying the BasicMeta metaclass makes me a bit nervous.

asmeurer commented:
I think that is all of my comments for now. My main concern here is the way the literal mapping works, it has a global mapping which never gets cleared. Each ask computation should have its own mapping (perhaps one starting from the base known_facts mapping). 

I also couldn't tell from the code if the literal mapping is reused in satisfiable() or if it makes a new mapping.

ShubhamKJha commented:
> 
> 
> 
> I see. Yes, we want to remove that. It is a lot of work though (see #13059). This is how I changed sympify in that PR: https://github.com/sympy/sympy/pull/13059/files#diff-50fcb76dd49035ed7d6ce34929ba5e0b. How does your change compare to that? Modifying the BasicMeta metaclass makes me a bit nervous.

`isinstance()` was costly hence I added `__sympy__` property to all the classes during `BasicMeta`. At that time I saw some performance gain of this approach and I think `sympify` is used at many places and many times even small gains should be significant.

ShubhamKJha commented:
> 
> 
> I think that is all of my comments for now. My main concern here is the way the literal mapping works, it has a global mapping which never gets cleared. Each ask computation should have its own mapping (perhaps one starting from the base known_facts mapping).
> 
> I also couldn't tell from the code if the literal mapping is reused in satisfiable() or if it makes a new mapping.

Currently, the mapping is actually encoding SymPy objects used to initialize the literals ( an attempt to reduce the use of SymPy objects). And the real encoding happens (again) in EncodedCNF, which is used by satisfiable.

asmeurer commented:
Can we merge the two encodings so it only happens once? That would also fix the global encodings problem, I think.

ShubhamKJha committed:
Modified the code for ask_generated to produce sorted version of precompiled CNF object

ShubhamKJha committed:
Fixed sorting in literals

ShubhamKJha committed:
Revert some changes since zip_longest is not available in Python 2

ShubhamKJha committed:
Removed literal_store in order to avoid memory leak and reduced encoding of Literals happening in the constructor itself. Instead encoding now happens at EncodedCNF only

ShubhamKJha committed:
Removed unnecessary `__lt__` function from `Literal`

ShubhamKJha commented:
> 
> 
> Can we merge the two encodings so it only happens once? That would also fix the global encodings problem, I think.

Placing encoding on CNF object appeared to be difficult since I had implemented interactions among different CNF objects. That would have required updating the encoding every such time. Currently, I have changed `Literal` to be a wrapper around `expressions` and the encoding happens at `EncodedCNF` only.

`Literal` using expressions certainly means some performance regression froms `__hash__` and `__eq__` functions. But it is not much in the test suite. We can improve over this in further PRs. It would need some refactoring of CNF class also.

ShubhamKJha commented:
ping @asmeurer

oscarbenjamin commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,373 @@
+        return Literal(expr)
+
+    klass = type(expr).__name__
+    if klass == 'Not':
```
Why are you using the type name string rather than isinstance?

ShubhamKJha replied:
It's changed. This is outdated.

oscarbenjamin replied:
So it is. Sorry!

oscarbenjamin commented on sympy/assumptions/cnf.py:
```diff
@@ -0,0 +1,347 @@
+"""
+The classes used here are for the internal use of assumptions system
+only and should not be used anywhere else as these don't posses the
```
posess

oscarbenjamin replied:
Whoops. Possess

ShubhamKJha committed:
Fix some minor mistake

ShubhamKJha committed:
Fix some minor mistakes

ShubhamKJha committed:
Merge branch 'speed_up_cnf_3' of https://github.com/ShubhamKJha/sympy into speed_up_cnf_3

ShubhamKJha commented:
ping @asmeurer, @oscarbenjamin, @jksuom does it need further edits? 

Any other features (including a good strategy for the early encoding of literals) can be implemented in future PRs, since that will require some refactoring of code at various levels.

jksuom commented on sympy/assumptions/ask.py:
```diff
@@ -1211,6 +1213,30 @@ def _extract_facts(expr, symbol, check_reversed_rel=True):
+                    args.append(None)
+                continue
+
+        if args and all(x is not None for x in args):
```
Perhaps `None not in args`.

oscarbenjamin commented on sympy/assumptions/ask.py:
```diff
@@ -1211,6 +1213,30 @@ def _extract_facts(expr, symbol, check_reversed_rel=True):
+        else:
+            facts.append(None)
+    args = [arg for arg in facts if arg is not None]
+    return CNF(set(args))
```
You can use a set comprehension directly rather than create a list and then turn it into a set: `{arg for arg in facts if arg is not None}` but... Why are you appending None to facts given that you then filter all Nones from facts?

It seems facts is defined as an empty list, then either has a frozenset appended or None (which is then filtered out). I'm not sure that the facts variable is needed at all.

I think that probably this whole function can be simplified substantially.

oscarbenjamin replied:
Oh sorry, I misread the indentation of the loops. Forget what I said above

ShubhamKJha replied:
Thanks. I simplified it.

ShubhamKJha replied:
> Oh sorry, I misread the indentation of the loops. Forget what I said above

I don't know but your opinion seems genuine. I have changed the code and it seems fine.

ShubhamKJha replied:
> It seems facts is defined as an empty list, then either has a frozenset appended or None (which is then filtered out). I'm not sure that the facts variable is needed at all.

Ok I understand now

oscarbenjamin replied:
Maybe this:
```python
def _extract_all_facts(expr, symbol):
    if isinstance(symbol, Relational):
        symbols = (symbol, symbol.reversed)
    else:
        symbols = (symbol,)

    facts = set()
    for clause in expr.clauses:
        args = []
        for literal in clause:
            if isinstance(literal.lit, AppliedPredicate):
                if literal.lit.arg not in symbols:
                    break
                args.append(Literal(literal.lit.func, literal.is_Not))
        else:
            if args:
                facts.add(frozenset(args))

    return CNF(facts)
```
Note that the break here gives an early exit from the literal loop. There should be a meaningful comment explaining why that happens and no fact is added in that case...

ShubhamKJha replied:
I had the exact changes. Will add a comment.

asmeurer commented:
> Placing encoding on CNF object appeared to be difficult since I had implemented interactions among different CNF objects. That would have required updating the encoding every such time. Currently, I have changed Literal to be a wrapper around expressions and the encoding happens at EncodedCNF only.

I expect that might even be faster than doing `__eq__` on large SymPy expressions. But we can see how this performs.

ShubhamKJha committed:
Simplified the code in _extract_all_facts from ask module

ShubhamKJha commented:
ping @asmeurer @oscarbenjamin @jksuom is it now good enough to merge?

jksuom commented on sympy/assumptions/ask.py:
```diff
@@ -1257,54 +1281,58 @@ def ask(proposition, assumptions=True, context=global_assumptions):
+    # using satask (still costly)
+    res = satask(proposition, assumptions=assumptions, context=context)
+
+    if res is not None:
```
It this test needed? It looks like `res` will be returned anyway on the next line.

ShubhamKJha replied:
Thanks for pointing out. Didn't realize.

jksuom commented:
This looks good to me.

ShubhamKJha committed:
Some minor fixes

asmeurer commented on sympy/assumptions/ask.py:
```diff
@@ -1481,7 +1517,7 @@ def get_known_facts():
         Equivalent(Q.even | Q.odd, Q.integer),
         Implies(Q.even, ~Q.odd),
-        Equivalent(Q.prime, Q.integer & Q.positive & ~Q.composite),
+        Implies(Q.prime, Q.integer & Q.positive & ~Q.composite),
```
Why was this changed?

ShubhamKJha replied:
Because `1` is neither prime nor composite. `satask` was giving the wrong answer for prime.

asmeurer replied:
Ah. I thought this was already well tested.

asmeurer replied:
Add a test for this.

ShubhamKJha replied:
> Ah. I thought this was already well tested.

I figured it out because for sometimes I tested ask with `_eval_ask` functions performed after `satask`. This was the only error that popped.


> Add a test for this.

Sure

ShubhamKJha committed:
Updated code in sathandlers and added tests for satask queries related to prime and composite

ShubhamKJha commented:
The failing test seems to be unrelated to the PR here. Can anyone restart tests?

jksuom commented:
Restart does not help. It seems that other pulls are also failing with an error in lambdify.

ShubhamKJha commented:
Ok, the tests are passing. Ping @jksuom @oscarbenjamin, this can be merged now.

jksuom commented:
Thanks, I also think that this can be merged.

oscargus commented:
Thanks @ShubhamKJha (and@rlamy )! Great to see some speed-ups!

(@ShubhamKJha  if you are not bored with this topic yet, feel free to remove the slow-marks for the no longer slow tests in another PR. These are probably the candidates.

```
537 test_negative - Took 0.450 seconds
538 test_zero - Took 0.502 seconds
539 test_positive - Took 0.575 seconds
540 test_bounded1 - Took 0.636 seconds
541 test_hermitian - Took 0.711 seconds
542 test_pow4 - Took 0.809 seconds
543 test_bounded3 - Took 0.839 seconds
544 test_real_pow - Took 0.994 seconds
545 test_pow2 - Took 1.036 seconds
546 test_even_query - Took 1.059 seconds
547 test_pow3 - Took 1.185 seconds
548 test_bounded2a - Took 1.331 seconds
549 test_odd_query - Took 1.628 seconds
550 test_imaginary - Took 1.676 seconds
551 test_bounded2b - Took 2.334 seconds
553 test_fidelity - Took 4.954 seconds
```
)

asmeurer commented:
Ah yes that definitely needs to be done, I forgot about that.

oscargus commented:
slow-markers are removed in #17367
